File: backend/api/models/graph.py
from pydantic import BaseModel, Field
from typing import Optional


class GraphNode(BaseModel):
    id: str
    name: str
    type: str
    layer: str
    description: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    columns: dict = Field(default_factory=dict)
    compiled_sql: Optional[str] = None
    raw_sql: Optional[str] = None
    depends_on: list[str] = Field(default_factory=list)
    referenced_by: list[str] = Field(default_factory=list)
    test_count: int = 0
    test_results: list[dict] = Field(default_factory=list)
    materialization: str = "view"
    schema_name: Optional[str] = None
    database: Optional[str] = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    column_count: int = 0


class GraphData(BaseModel):
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    stats: dict = Field(default_factory=dict)

File: frontend/src/types/graph.ts
export interface GraphNode {
  id: string
  name: string
  type: string
  layer: string
  description?: string
  tags: string[]
  columns: Record<string, ColumnDef>
  compiled_sql?: string
  raw_sql?: string
  depends_on: string[]
  referenced_by: string[]
  test_count: number
  test_results: TestResult[]
  materialization: string
  schema_name?: string
  database?: string
}

export interface ColumnDef {
  name?: string
  description?: string
  data_type?: string
  meta?: Record<string, unknown>
  tags?: string[]
}

export interface TestResult {
  name: string
  status: 'pass' | 'fail' | 'warn'
  column: string
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  column_count: number
}

export interface GraphStats {
  total_models: number
  total_sources: number
  total_exposures: number
  total_tests: number
  test_coverage_pct: number
  models_by_layer: Record<string, number>
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
  stats: GraphStats
}
